apps/mobile/src/state/contacts/atom/submitContactsActionAtom.ts

```diff
@@ -44,90 +50,142 @@ export const submitContactsActionAtom = atom(
+            one.flags.imported &&
+            !numbersToImport.includes(one.computedValues.normalizedNumber)
         )
-      }),
-      TE.match(
-        (e) => {
-          if (e._tag === 'InitialImportContactsQuotaReachedError') {
-            Alert.alert(t('contacts.initialImportContactsQuotaReachedError'))
-            return false
-          }
-          if (e._tag === 'ImportContactsQuotaReachedError') {
-            Alert.alert(t('contacts.importContactsQuotaReachedError'))
-            return false
-          }
-          if (e._tag !== 'NetworkError') {
-            reportError('error', new Error('error while submitting contacts'), {
-              e,
-            })
-          }
+      )
 
-          Alert.alert(toCommonErrorMessage(e, t) ?? t('common.unknownError'))
-          return false
-        },
-        (contacts) => {
-          const importedNumbers = new Set(
-            contacts.map((one) => one.computedValues.normalizedNumber)
+      const contactsThatShouldBeImported = numbersToImport
+        .map((oneNumberToImport): StoredContactWithComputedValues | undefined =>
+          allContacts.find(
+            (oneContact) =>
+              oneContact.computedValues.normalizedNumber === oneNumberToImport
           )
+        )
+        .filter(notEmpty)
 
-          set(storedContactsAtom, (storedContact) =>
-            storedContact.map((oneContact) =>
-              oneContact.computedValues &&
-              importedNumbers.has(oneContact.computedValues?.normalizedNumber)
-                ? {
-                    ...oneContact,
-                    flags: {
-                      ...oneContact.flags,
-                      imported: true,
-                    },
-                  }
-                : {...oneContact, flags: {...oneContact.flags, imported: false}}
-            )
-          )
+      const newContactsToImport =
+        // IF there are no contacts to remove, we can do an incremental import
+        HashSet.size(contactsThatShouldBeRemovedFromImport) > 0
+          ? contactsThatShouldBeImported
+          : contactsThatShouldBeImported.filter((one) => !one.flags.imported)
+
+      const importedNumbersSoFarRef = yield* _(
+        Ref.make(HashSet.empty<E164PhoneNumber>())
+      )
+      yield* _(
+        pipe(
+          newContactsToImport,
+          Array.chunksOf(CONTACT_IMPORT_BATCHES),
+          Array.map((chunkToImport, i) => {
+            const replace =
+              i === 0 && HashSet.size(contactsThatShouldBeRemovedFromImport) > 0
+            return contactApi
+              .importContacts({
+                body: {
+                  contacts: chunkToImport.map((one) => one.computedValues.hash),
+                  replace,
+                },
+              })
+              .pipe(
+                Effect.zipLeft(
+                  Ref.update(
+                    importedNumbersSoFarRef,
+                    HashSet.union(
+                      newContactsToImport.map(
```

Comment:
Shouldn't be here `chunkToImport` instead of `newContactsToImport`? Or maybe I don't understand it correctly ...

Reply:
You are right! Good catch!